detector/include/DCH_info.h

```diff
@@ -0,0 +1,341 @@
+    ff_check_positive_parameter(this->dch_first_sense_r,"radius of first layer cells" );
+
+
+    // intialize layer 1 from input parameters
+    {
+        DCH_info_layer layer1_info;
+        layer1_info.layer         = 1;
+        layer1_info.nwires        = 2*this->dch_ncell0;
+        layer1_info.height_z0     = dch_first_width;
+        layer1_info.radius_sw_z0  = dch_first_sense_r;
+        layer1_info.radius_fdw_z0 = dch_first_sense_r - 0.5*dch_first_width;
+        layer1_info.radius_fuw_z0 = dch_first_sense_r + 0.5*dch_first_width;
+        layer1_info.width_z0      = TMath::TwoPi()*dch_first_sense_r/this->dch_ncell0;
+
+        this->database.emplace(layer1_info.layer, layer1_info);
+    }
+
+    // some parameters of the following layer are calculated based on the previous ones
+    // the rest are left as methods of DCH_info or DCH_info_layer class
+    // loop over all layers, skipping the first one
+    for(int ilayer = 2; ilayer<= this->dch_nlayers; ++ilayer)
+    {
+        // initialize here an object that will contain one row from the excel table
+        DCH_info_layer layer_info;
+
+        // the loop counter actually corresponds to the layer number
+        layer_info.layer = ilayer;
+        // nwires is twice the number of cells in this particular layer (ilayer)
+        layer_info.nwires = 2*(this->dch_ncell0 + this->dch_ncell_increment*Get_nsuperlayer_minus_1(ilayer) );
+
+        // the previous layer info is needed to calculate parameters of current layer
+        auto previousLayer = this->database.at(ilayer-1);
+
+        //calculate height_z0, radius_sw_z0
+        {
+            double h  = previousLayer.height_z0;
+            double ru = previousLayer.radius_fuw_z0;
+            double rd = previousLayer.radius_fdw_z0;
+
+            if(0 == Get_nsuperlayer_minus_1(ilayer))
+                layer_info.height_z0 = h*ru/rd;
+            else
+                layer_info.height_z0 = TMath::TwoPi()*ru/(0.5*layer_info.nwires - TMath::Pi());
+
+            layer_info.radius_sw_z0 = 0.5*layer_info.height_z0 + ru;
+        }
+
+        //calculate radius_fdw_z0, radius_fuw_z0, width_z0
+        layer_info.radius_fdw_z0 = previousLayer.radius_fuw_z0;
+        layer_info.radius_fuw_z0 = previousLayer.radius_fuw_z0 + layer_info.height_z0;
+        layer_info.width_z0 = TMath::TwoPi()*layer_info.radius_sw_z0/(0.5*layer_info.nwires);
+
+        // according to excel, width_z0 == height_z0
+        if(fabs(layer_info.width_z0 - layer_info.height_z0)>1e-4)
+            throw std::runtime_error("fabs(l.width_z0 - l.height_z0)>1e-4");
+
+
+
+        this->database.emplace(ilayer, layer_info);
+    }
+
+    std::cout << "\t+ Total size of DCH database = " << database.size() << std::endl;
+    return;
+}
+
+void DCH_info_struct::Show_DCH_info_database(std::ostream & oss) const
+{
+    if( this->IsDatabaseEmpty() )
+        throw std::runtime_error("Can not show empty database!");
```

Comment:
Just print `Database empty` after the global parameters, instead of throwing an error here. Maybe people just want to see the global parameters.